### ISSUE TITLE
define custom onboarding flow for azure with sources-api

### ISSUE THREAD

**Issue by infinitewarp** [cloudigrade/cloudigrade#1026]
## Summary

As a swatch customer, I want cloudigrade to have access to my Azure account following my interactions with sources.

Generally, we want the following sequence of events:
- customer comes to sources UI/API and starts defining an Azure application/source
- sources-api retrieves the Azure policy from cloudigrade
  - this is required if we don't have a truly public API for the policy.
  - …or sources-api just retrieves the URL path to our policy if the path is publicly accessible
- sources-api makes a special request [probably using this Azure API](https://docs.microsoft.com/en-us/azure/lighthouse/how-to/onboard-customer#azure-cli) that includes cloudigrade's policy and other customer-provided credentials
- sources-api monitors that API call for completion before saving the source; this could take several minutes.
- sources-api saves its source, and the normal flow proceeds to notify via kafka
- cloudigrade sees the kafka message and sets up CloudAccount, etc. objects.

## Acceptance Criteria

- [x] Verify when following the above process via sources-api interaction, cloudigrade receives and handles kafka messages:
  - [x] create AzureCloudAccount, etc.
  - [x] delete AzureCloudAccount, etc. and all related objects (runs, etc.)
- [x] Verify that cloudigrade checks for access to provided subscription id, and appropriately updates the availability check.

## Assumptions and Questions

- @gtanzillo has a monthly chat with sources folks (Drew?) and will start the conversation on 2021-10-11.
- Much of this is having a conversation with sources devs and convincing them that this is the way.
  - This feels a lot like the superkey flow, and that's why it feels like it would be okay to make sources-api's responsibility.
- See https://github.com/cloudigrade/cloudigrade/pull/991 which added a template
- We may want to change the path defined by #991 but that is not strictly required
- We may want to work with 3scale config to allow the specific path for the Azure policy to be unauthenticated.
- No availability check support for the Azure account yet. At this point, we only know that the account *exists* but we haven't communicated anything yet.
- Is it possible to get an "update" event from sources-api in this flow?
  - We suspect *no*. It should not be possible to replace the subscription ID for the existing offering. It should instead be a process of deleting the old and creating a new.

## Historic update notes

Update 2021-12-08: Per https://github.com/cloudigrade/cloudigrade/issues/1026#issuecomment-988257531 we expect this issue is no longer blocked.

~This issue is blocked on sources-api implementing their first half of this issue. The "crawl" phase appears to be partially done and ready in stage, but the implementation may be incomplete. See https://issues.redhat.com/browse/RHCLOUD-16681 and https://github.com/RedHatInsights/sources-superkey-worker/pull/42~
[GT] Update 2021-12-07: The above is no longer relevant because super key cannot be used with lighthouse and manual mode is required.

Update 2021-11-24: still waiting on https://issues.redhat.com/browse/RHCLOUD-12576 That issue specifically gets the subscription ID from the customer during setup and sends that ID to cloudigrade so that we can cross-reference our availability/access in Azure.

[GT] Update 2021-12-07: Had a discussion with Drew and he confirmed that we actually have everything we need to move forward. See comment [here in RHCLOUD-12576](https://issues.redhat.com/browse/RHCLOUD-12576?focusedCommentId=19428194&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19428194)

**Comment by katherine-black** [comment]
Tracking https://issues.redhat.com/browse/RHCLOUD-16927

**Comment by katherine-black** [comment]
Looks like sources may be ready? https://issues.redhat.com/browse/RHCLOUD-12576?focusedCommentId=19428194&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19428194

**Comment by lindgrenj6** [comment]
@katherine-black yep we are ready on the backend (sorry, meant to do this sooner but had a bunch of PTO and we moved) - if you want to simulate what the UI will be sending (when it's done) it'll be a call like this:

```
POST https://console.stage.redhat.com/api/sources/v3.1/bulk_create
{
	"sources": [
		{
			"name": "lighthousetestsource",
			"source_type_name": "azure"
		}
	],
	"authentications": [
		{
			"resource_type": "application",
			"resource_name": "cloud-meter",
			"username": "your-subscription-id",
			"authtype": "lighthouse_subscription_id"
		}
	],
	"applications": [
		{
			"source_name": "lighthousetestsource",
			"application_type_name": "cloud-meter"
		}
	]
}
```

Just replace the username there with whatever subscription id was ran with the lighthouse button.

**Comment by katherine-black** [comment]
Here is a run-through of this integration as done in the stage environment. 

First, I grab the offer link being served by sources, currently the UI is unavailable for this, for I use the following command to fetch it.

```
❯ http https://console.stage.redhat.com/api/sources/v3.1/app_meta_data | jq -r '.data[] | select(.name=="azure_lighthouse_template") | .payload'

https://portal.azure.com/#create/Microsoft.Template/uri/https%3A%2F%2Fcloudigrade-stage-azure-offer-cloudigrade-stage.apps.crcs02ue1.urby.p1.openshiftapps.com%2Fapi%2Fcloudigrade%2Fv2%2Fazure-offer-template%2F
```

After following the above link, I register the rh dir2 account to be managed by our stage account. The deployment succeeds.
![Screen Shot 2022-01-11 at 13 03 17](https://user-images.githubusercontent.com/34514071/149024736-c400f0a9-7930-4434-ae01-61a751ad79b5.png)

Next, I manually create the sources object, (once again, I'm told UI for this is coming)
```
echo -n '{"sources":[{"name":"lighthousetestsource","source_type_name":"azure"}],"authentications":[{"resource_type":"application","resource_name":"cloud-meter","username":"23341ae3-be0b-4b80-b212-85d03f37d607","authtype":"lighthouse_subscription_id"}],"applications":[{"source_name":"lighthousetestsource","application_type_name":"cloud-meter"}]}' | http https://console.stage.redhat.com/api/sources/v3.1/bulk_create

HTTP/1.1 201 Created
Cache-Control: max-age=0, private, must-revalidate
Connection: keep-alive
Content-Length: 678
Content-Type: application/json; charset=utf-8
Date: Tue, 11 Jan 2022 21:26:43 GMT
ETag: W/"41c28518bcb9d578afa0d16a76766bbc"
Server: openresty
Set-Cookie: 8f585a7d113e400d14faa100d81cbdaf=def9effde4f45adb62c1aa389b84a4b4; path=/; HttpOnly; Secure; SameSite=None
Vary: Origin
X-Frame-Options: SAMEORIGIN
x-content-type-options: nosniff
x-request-id: abd7c508-4711-4729-8eaa-25fc11beffd4
x-rh-edge-cache-status: Miss from child, Miss from parent
x-rh-edge-reference-id: 0.5e8cd017.1641936403.24fe
x-rh-edge-request-id: 24fe
x-rh-insights-request-id: 2a567bf703ca4ecd9f7092a92d4b0f09
x-runtime: 0.602296

{
    "applications": [
        {
            "application_type_id": "3",
            "created_at": "2022-01-11T21:26:42Z",
            "extra": {},
            "id": "97948",
            "source_id": "107222",
            "tenant": "6089719",
            "updated_at": "2022-01-11T21:26:42Z"
        }
    ],
    "authentications": [
        {
            "authtype": "lighthouse_subscription_id",
            "id": "95999",
            "resource_id": "97948",
            "resource_type": "Application",
            "source_id": "107222",
            "tenant": "6089719",
            "username": "23341ae3-be0b-4b80-b212-85d03f37d607"
        }
    ],
    "endpoints": null,
    "sources": [
        {
            "app_creation_workflow": "manual_configuration",
            "created_at": "2022-01-11T21:26:42Z",
            "id": "107222",
            "name": "lighthousetestsource",
            "source_type_id": "3",
            "tenant": "6089719",
            "uid": "adb74e1d-569d-4b65-8daf-59229890d843",
            "updated_at": "2022-01-11T21:26:42Z"
        }
    ]
}
```

Give it a few moments and you'll have...

```
❯ http https://console.stage.redhat.com/api/sources/v3.1/sources/107222
HTTP/1.1 200 OK
Cache-Control: max-age=0, private, must-revalidate
Connection: keep-alive
Content-Length: 405
Content-Type: application/json; charset=utf-8
Date: Tue, 11 Jan 2022 21:28:59 GMT
ETag: W/"08bfb6dc26700dccd91ec0bbebc2b6e7"
Server: openresty
Set-Cookie: 8f585a7d113e400d14faa100d81cbdaf=a733d3c0fbc7dea0b9d2a3215d615c2c; path=/; HttpOnly; Secure; SameSite=None
Vary: Origin
X-Frame-Options: SAMEORIGIN
x-content-type-options: nosniff
x-request-id: d8dc0311-9a37-4851-9ec9-dc44c54d05d4
x-rh-edge-cache-status: Miss from child, Miss from parent
x-rh-edge-reference-id: 0.642ec417.1641936539.18ac837f
x-rh-edge-request-id: 18ac837f
x-rh-insights-request-id: 130b8864c4ed44348fd90e7ea008c188
x-runtime: 0.009974

{
    "app_creation_workflow": "manual_configuration",
    "availability_status": "available",
    "created_at": "2022-01-11T21:26:42.578419Z",
    "id": "107222",
    "last_available_at": "0001-01-01T00:00:00Z",
    "last_checked_at": "2022-01-11T21:27:02.724303Z",
    "name": "lighthousetestsource",
    "paused_at": "0001-01-01T00:00:00Z",
    "source_type_id": "3",
    "uid": "adb74e1d-569d-4b65-8daf-59229890d843",
    "updated_at": "2022-01-11T21:27:02.742373Z"
}

❯ http https://cloud.stage.redhat.com/api/cloudigrade/v2/accounts/16957/
HTTP/1.1 200 OK
Allow: GET, HEAD, OPTIONS
Cache-Control: private
Connection: keep-alive
Content-Length: 483
Content-Type: application/json
Date: Tue, 11 Jan 2022 21:28:37 GMT
Server: openresty
Set-Cookie: 8f585a7d113e400d14faa100d81cbdaf=a733d3c0fbc7dea0b9d2a3215d615c2c; path=/; HttpOnly; Secure; SameSite=None
referrer-policy: same-origin
x-cloudigrade-request-id: 10576233-c7d0-4ca6-8795-cff7be368d6b
x-content-type-options: nosniff
x-frame-options: DENY
x-rh-edge-cache-status: Miss from child, Miss from parent
x-rh-edge-reference-id: 0.5d2ec417.1641936517.41b8526d
x-rh-edge-request-id: 41b8526d
x-rh-insights-request-id: 579d614db3ad45a7b22104a9b48fe8a4

{
    "account_id": 16957,
    "cloud_type": "azure",
    "content_object": {
        "azure_cloud_account_id": 4,
        "created_at": "2022-01-11T21:27:00.699307Z",
        "subscription_id": "23341ae3-be0b-4b80-b212-85d03f37d607",
        "updated_at": "2022-01-11T21:27:00.699333Z"
    },
    "created_at": "2022-01-11T21:27:00.705329Z",
    "is_enabled": true,
    "platform_application_id": 97948,
    "platform_application_is_paused": false,
    "platform_authentication_id": 95999,
    "platform_source_id": 107222,
    "updated_at": "2022-01-11T21:27:00.739057Z",
    "user_id": 5603
}
```

The backend logs for the creation and destruction of the azure account are here: http://pastebin.test.redhat.com/1020485

**Comment by infinitewarp** [comment]
Looks like that covers our ACs. I'm happy to see forward progress on the Azure integration. Closing!
